File: machine learning models/Neural Networks based Insurance Purchase Prediction/Python/insurancenn.py
import csv
import numpy as np
import math
from scipy.optimize import minimize

# function that normalizes each predictor
def normalize(X):
    m = np.size(X, 0)
    # n = np.size(X, 1)
    maxval = np.max(X, 0)
    minval = np.min(X, 0)
    minMat = np.tile(minval, (m, 1))
    DifMat = np.tile(maxval, (m, 1)) - minMat
    Xnorm = np.divide(X - minMat, DifMat)
    return Xnorm

# function that return sigmoid of z
def sigmoid(z):
    g = 1 / (1 + np.exp(-z))
    return g

def gradient(z):
    g1 = np.column_stack((np.ones(np.size(z, 0)), sigmoid(z)))
    g2 = np.column_stack((np.ones(np.size(z, 0)), (1 - sigmoid(z))))
    g = np.multiply(g1, g2)
    return g

# cost function
def computeCost(theta, input_num, hidden_num, label_num, X, y, lamba):
    theta1 = theta[0:hidden_num * (1 + input_num)]
    theta2 = theta[hidden_num * (1 + input_num):len(theta)]
    Theta1 = theta1.reshape(hidden_num, 1 + input_num, order="F")
    Theta2 = theta2.reshape(label_num, 1 + hidden_num, order="F")
    m = np.size(X, 0)
    # convert the labels,1 to 10,2 to 01
    I = np.eye(2)
    Y = I[y - 1, :]

    A1 = np.column_stack((np.ones(m), X))
    Z2 = np.dot(A1, Theta1.T)
    A2 = np.column_stack((np.ones(np.size(Z2, 0)), sigmoid(Z2)))
    Z3 = np.dot(A2, Theta2.T)
    H = sigmoid(Z3)
    # A3 = H                      # never used

    # Feedforward & cost function
    J = np.sum(np.multiply(- Y, np.log(H)) +
               np.multiply(- (1 - Y), np.log(1 - H))) / m

    # Regularization
    reg = (lamba / (2*m)) * (np.sum(Theta1[:, 1:np.size(Theta1, 1)]**2) +
                             np.sum(Theta2[:, 1:np.size(Theta2, 1)]**2))

    J = J + reg
    return J

# cost function
def cost_grad(theta, input_num, hidden_num, label_num, X, y, lamba):
    theta1 = theta[0:hidden_num * (1 + input_num)]
    theta2 = theta[hidden_num * (1 + input_num):len(theta)]
    Theta1 = theta1.reshape(hidden_num, 1 + input_num, order="F")
    Theta2 = theta2.reshape(label_num, 1 + hidden_num, order="F")
    m = np.size(X, 0)
    # convert the labels,1 to 10,2 to 01
    I = np.eye(2)
    Y = I[y - 1, :]

    A1 = np.column_stack((np.ones(m), X))
    Z2 = np.dot(A1, Theta1.T)
    A2 = np.column_stack((np.ones(np.size(Z2, 0)), sigmoid(Z2)))
    Z3 = np.dot(A2, Theta2.T)
    H = sigmoid(Z3)
    A3 = H

    # Backpropagation & Big deltas
    delta3 = A3 - Y
    delta2 = np.multiply(np.dot(delta3, Theta2), gradient(Z2))
    delta2 = delta2[:, 1:np.size(delta2, 1)]

    # Accumulate gradients
    BigDelta1 = np.dot(delta2.T, A1)
    BigDelta2 = np.dot(delta3.T, A2)

    # Regularized Gradient
    Theta1_grad = BigDelta1 / m + (lamba / m) *\
        np.column_stack((np.zeros(np.size(Theta1, 0)),
                         Theta1[:, 1:np.size(Theta1, 1)]))

    Theta2_grad = BigDelta2 / m + (lamba / m) *\
        np.column_stack((np.zeros(np.size(Theta2, 0)),
                         Theta2[:, 1:np.size(Theta2, 1)]))

    grad = np.concatenate(
        (Theta1_grad.reshape(hidden_num * (1 + input_num), 1, order="F"),
         Theta2_grad.reshape(label_num * (1 + hidden_num), 1, order="F")),
        axis=0)
    grad = np.squeeze(grad)
    return grad


def predict(Theta1, Theta2, X):
    m = np.size(X, 0)
    h1 = sigmoid(np.dot(np.column_stack((np.ones(m), X)), Theta1.T))
    h2 = sigmoid(np.dot(np.column_stack((np.ones(m), h1)), Theta2.T))
    pred = np.argmax(h2, 1) + 1
    return pred

	
# read data from csv file
datPath = r"C:\xxx\insurance.csv"
with open(datPath, 'rb') as csvfile:
    csvData = csv.reader(csvfile)
    datList = []
    for row in csvData:
        datList.append(row)

# get the colnames in the first row and remove it
colNames = datList.pop(0)

# convert list to matrix
data = np.array(datList)

# setting up
input_num = 21
hidden_num = 100
label_num = 2
train_pc = 0.66

# insurance data set
y = np.array(data[:, 30], "str")

# index of X selected
idx = np.array([i for j in (range(7, 14), range(15, 29)) for i in j])
X = np.array(data[:, idx], "float")

# normalize X
X = normalize(X)
m = np.size(X, 0)
n = np.size(X, 1)

# y is a 1d array
y = np.array([{"Yes": 2, "No": 1}[y[i]] for i in range(len(y))])

# Stratified sampling
size1 = math.floor(train_pc*sum(y == 1))
size2 = math.floor(train_pc*sum(y == 2))
idxY1 = np.linspace(0, len(y) - 1, len(y), dtype="int")[y == 1]
idxY2 = np.linspace(0, len(y) - 1, len(y), dtype="int")[y == 2]
B1train = np.concatenate((np.random.choice(idxY1, size1, False),
                          (np.random.choice(idxY2, size2, False))), axis=0)
yTrain = y[B1train]
yTest = np.delete(y, B1train, axis=0)
XTrain = X[B1train, :]
XTest = np.delete(X, B1train, axis=0)

# lambda and initial value
lamba = 0.05
epsilon = 0.1
Theta1 = np.random.rand(hidden_num, 1 + input_num) * 2 * epsilon - epsilon
Theta2 = np.random.rand(label_num, 1 + hidden_num) * 2 * epsilon - epsilon
theta = np.concatenate(
    (Theta1.reshape(hidden_num * (1 + input_num), 1, order="F"),
     Theta2.reshape(label_num * (1 + hidden_num), 1, order="F")))
theta = np.squeeze(theta)

# Optimization using scipy.optimize.minimize
res2 = minimize(computeCost, theta,
                args=(input_num, hidden_num, label_num, XTrain, yTrain, lamba),
                jac=cost_grad, options={"maxiter": 50})

theta1 = res2.x[0:hidden_num * (1 + input_num)]
theta2 = res2.x[hidden_num * (1 + input_num):len(theta)]
Theta1 = theta1.reshape(hidden_num, 1 + input_num, order="F")
Theta2 = theta2.reshape(label_num, 1 + hidden_num, order="F")

predTrain = predict(Theta1, Theta2, XTrain)
print('Training set Accuracy: %f\n' % (np.mean(predTrain == yTrain)*100))
predTest = predict(Theta1, Theta2, XTest)
print('Test set Accuracy: %f\n' % (np.mean(predTest == yTest)*100))

File: machine learning models/Neural Networks based Insurance Purchase Prediction/R/insurancenn.R
# cleaning
rm(list=ls())
path="C:/xxx"
setwd(path)
# use source if separated in another file

computeCost <- function(theta,input_num,hidden_num,label_num,X,y,lambda) {
  # making Theta1 and Theta2 matrices from the vector theta 
  # http://ufldl.stanford.edu/wiki/index.php/Backpropagation_Algorithm
  Theta1 <- matrix(theta[1:hidden_num*(input_num+1)],hidden_num,(input_num+1))
  Theta2 <- matrix(theta[-1:-(hidden_num*(input_num+1))],label_num,(hidden_num+1))
  m <- dim(X)[1]
  
  # convert the labels,1 to 10,2 to 01
  I <- diag(label_num)
  Y <- matrix(0,m,label_num)
  for (i in (1:m)) {
    Y[i,] = I[y[i],]
  }
  
  A1 <- cbind(rep(1,m),X) # A1: m x (input_num+1)
  Z2 <- A1%*%t(Theta1) # Z2: m x hidden_num; Theta1: hidden_num x (input_num+1)
  A2 <- cbind(rep(1,m),sigmoid(Z2)) # A2: m x (hidden_num+1)
  Z3 <- A2%*%t(Theta2)# Z3: m x number of output nodes; Theta2: number of output nodes x (number of hidden nodes+1)
  H <- sigmoid(Z3) # A3 and H: m x number of output nodes
  A3 <- H
  
  # Feedforward & cost function
  J = (1/m)*sum((-Y)*log(H)-(1-Y)*log(1-H))
  
  # Regularization 
  reg = (lambda/(2))*(sum(Theta1[,-1]^2)+sum(Theta2[,-1]^2))
  J = J+reg
  
  return(J)
}

computeGradient <- function(theta,input_num,hidden_num,label_num,X,y,lambda) {
  # making Theta1 and Theta2 matrices from the vector theta 
  # http://ufldl.stanford.edu/wiki/index.php/Backpropagation_Algorithm
  Theta1 <- matrix(theta[1:hidden_num*(input_num+1)],hidden_num,(input_num+1))
  Theta2 <- matrix(theta[-1:-(hidden_num*(input_num+1))],label_num,(hidden_num+1))
  m <- dim(X)[1]
  
  # convert the labels,1 to 10,2 to 01
  I <- diag(label_num)
  Y <- matrix(0,m,label_num)
  for (i in (1:m)) {
    Y[i,] = I[y[i],]
  }
  
  A1 <- cbind(rep(1,m),X) # A1: m x (input_num+1)
  Z2 <- A1%*%t(Theta1) # Z2: m x hidden_num; Theta1: hidden_num x (input_num+1)
  A2 <- cbind(rep(1,m),sigmoid(Z2)) # A2: m x (hidden_num+1)
  Z3 <- A2%*%t(Theta2)# Z3: m x number of output nodes; Theta2: number of output nodes x (number of hidden nodes+1)
  H <- sigmoid(Z3) # A3 and H: m x number of output nodes
  A3 <- H
  
  # Feedforward & cost function
  J = (1/m)*sum((-Y)*log(H)-(1-Y)*log(1-H))
  
  # Regularization 
  reg = (lambda/(2))*(sum(Theta1[,-1]^2)+sum(Theta2[,-1]^2))
  J = J+reg
  # Backpropagation & Big deltas
  delta3 = A3-Y # m x label_num
  
  delta2 = (delta3%*%Theta2)*gradient(Z2) 
  # delta3: m x label_num, Theta2: label_num x (hidden_num+1), gradient(Z2): a2 * (1-a2): m x (hidden_num+1)
  # delta2: m x (hidden_num+1)
  delta2 = delta2[,-1] # now delta2: m x hidden_num
  BigDelta1 = t(delta2)%*%A1 # ((hidden_num) x m) x (m x (input_num+1)) = ((hidden_num) x (input_num+1))
  BigDelta2 = t(delta3)%*%A2 # ((label_num) x m) x (m x (hidden_num+1)) = (label_num) x (hidden_num+1)
  
  # Regularized Gradient
  Theta1_grad = BigDelta1/m+lambda*cbind(rep(0,hidden_num),Theta1[,-1]) # Theta1: hidden_num x (input_num+1)
  Theta2_grad = BigDelta2/m+lambda*cbind(rep(0,label_num),Theta2[,-1]) # Theta2: label_num x (hidden_num+1)
  
  # Making vector theta from Theta1 and Theta2 to return
  grad = c(c(Theta1_grad),c(Theta2_grad))
  
  return(grad)
}

gradient<-function (z) {
  p <- dim(z)[1]
  a <- cbind(rep(p,1),sigmoid(z))
  g <- a * (1-a)
}

normalize <- function(X) {
  m <- dim(X)[1]
  n <- dim(X)[2]
  maxval <- apply(X,2,max)
  minval <- apply(X,2,min)
  
  # example here: How to divide each row of a matrix by elements of a vector in R
  # http://stackoverflow.com/questions/20596433/how-to-divide-each-row-of-a-matrix-by-elements-of-a-vector-in-rhttp://stackoverflow.com/questions/20596433/how-to-divide-each-row-of-a-matrix-by-elements-of-a-vector-in-rhttp://stackoverflow.com/questions/20596433/how-to-divide-each-row-of-a-matrix-by-elements-of-a-vector-in-r
  # mat<-matrix(1,ncol=2,nrow=2,TRUE)
  # dev<-c(5,10)
  # mat/dev or mat-dev would apply dev on mat column-wise, to be row-wise do:
  # t(t(mat) / dev)
  
  Xnumerator <- t(t(X)-minval)
  Xnorm <- t(t(Xnumerator)/(maxval-minval))
}

predict <- function(Theta1,Theta2,X) {
  m <- dim(X)[1]
  n <- dim(X)[2]
  h1 <- sigmoid(cbind(rep(1,m),X)%*%t(Theta1))
  h2 <- sigmoid(cbind(rep(1,m),h1)%*%t(Theta2))
  pred <- apply(h2,1,function(x) which(x==max(x)))
}

sigmoid <- function(z) {
  g = 1/(1+exp(-z))
}

hidden_num <- 100
label_num <- 2 # 10 and 01       
train_pc <- .66

data <- read.csv('insurance.csv')

y <- (data[,31]=="Yes")+1 # 1 and 2
X <- data.matrix(cbind(data[,8:14],data[,16:29])) # only numeric fields, converted to matrix
X <- normalize(X) # binary fields are not affected from min-max normalization

input_num <- dim(X)[2]

max_instance <- max(sum(y==1),sum(y==2))

train_ind <- c(sample(which(y==1),floor(train_pc*sum(y==1)),replace = FALSE),sample(which(y==2),floor(train_pc*sum(y==2)),replace = FALSE))
yTrain <- y[train_ind]
yTest <- y[-train_ind]
xTrain <- X[train_ind,]
xTest <- X[-train_ind,]

lambda <- 0.05
epsilon <- 0.1
Theta1 <- matrix(runif(hidden_num*(1+input_num)),hidden_num,1+input_num)*2*epsilon-epsilon # random beween -epsilon and epsilon
Theta2 <- matrix(runif(label_num*(1+hidden_num)),label_num,1+hidden_num)*2*epsilon-epsilon # random beween -epsilon and epsilon

theta <- c(c(Theta1),c(Theta2))

# result <- optim(theta, fn=computeCost, gr=computeGradient, input_num = input_num,hidden_num = hidden_num,label_num = label_num,X = X,y = y,lambda = lambda,method="BFGS",hessian = TRUE, control=list(trace=TRUE))

result <- optim(theta, fn=computeCost, gr=computeGradient, input_num = input_num,hidden_num = hidden_num,label_num = label_num,X = X,y = y,lambda = lambda,method="BFGS",control=list(trace=TRUE,maxit=1000000))

theta <- result$par
  
Theta1 <- matrix(theta[1:hidden_num*(input_num+1)],hidden_num,(input_num+1))
Theta2 <- matrix(theta[-1:-(hidden_num*(input_num+1))],label_num,(hidden_num+1))

predTrain = predict(Theta1,Theta2,xTrain)
cat('Training Set Accuracy:',mean(predTrain == yTrain)*100)

predTest = predict(Theta1,Theta2,xTest)
cat('Test Set Accuracy:',mean(predTest == yTest)*100)

             